src/sync.rs

```diff
@@ -0,0 +1,351 @@
+    /// A rustc-pull has failed, probably a git operation error has occurred.
+    PullFailed(anyhow::Error),
+}
+
+impl From<anyhow::Error> for RustcPullError {
+    fn from(error: Error) -> Self {
+        Self::PullFailed(error)
+    }
+}
+
+pub struct PullResult {
+    pub merge_commit_message: String,
+}
+
+pub struct GitSync {
+    context: SyncContext,
+    proxy: JoshProxy,
+}
+
+impl GitSync {
+    pub fn new(context: SyncContext, proxy: JoshProxy) -> Self {
+        Self { context, proxy }
+    }
+
+    pub fn rustc_pull(&self) -> Result<PullResult, RustcPullError> {
+        // The upstream commit that we want to pull
+        let upstream_sha = {
+            let out = run_command([
+                "git",
+                "ls-remote",
+                &format!("https://github.com/{UPSTREAM_REPO}"),
+                "HEAD",
+            ])
+            .context("cannot fetch upstream commit")?;
+            out.split_whitespace()
+                .next()
+                .unwrap_or_else(|| panic!("Could not obtain Rust repo HEAD from remote: '{out}'"))
+                .to_owned()
+        };
+
+        ensure_clean_git_state();
+
+        // Make sure josh is running.
+        let josh = self
+            .proxy
+            .start(&self.context.config)
+            .context("cannot start josh-proxy")?;
+        let josh_url = josh.git_url(
+            UPSTREAM_REPO,
+            Some(&upstream_sha),
+            &self.context.config.construct_josh_filter(),
+        );
+
+        let orig_head = run_command(["git", "rev-parse", "HEAD"])?;
+        println!(
+            "previous upstream base: {:?}",
+            self.context.last_upstream_sha
+        );
+        println!("new upstream base: {upstream_sha}");
+        println!("original local HEAD: {orig_head}");
+
+        // If the upstream SHA hasn't changed from the latest sync, there is nothing to pull
+        // We distinguish this situation for tools that might not want to consider this to
+        // be an error.
+        if let Some(previous_base_commit) = self.context.last_upstream_sha.as_ref() {
+            if *previous_base_commit == upstream_sha {
+                return Err(RustcPullError::NothingToPull);
+            }
+        }
+
+        // Update the last upstream SHA file. As a separate commit, since making it part of
+        // the merge has confused the heck out of josh in the past.
+        // We pass `--no-verify` to avoid running git hooks.
+        // We do this before the merge so that if there are merge conflicts, we have
+        // the right rust-version file while resolving them.
+        std::fs::write(&self.context.last_upstream_sha_path, &upstream_sha).with_context(|| {
+            anyhow::anyhow!(
+                "cannot write upstream SHA to {}",
+                self.context.last_upstream_sha_path.display()
+            )
+        })?;
+
+        // Fetch the latest upstream HEAD so we can get a summary. Use the Josh URL for caching.
```

Comment:
This will import the entire rustc git history into my Miri checkout, won't it? Please don't, I'd rather not waste 1-2GB of disk space like that.  The rustc-push logic specifically has this RUSTC_GIT env var to avoid this exact situation.